File: 20-ValidParenthesis.js
/*
Single pass scan, using stack as memory

O(N) time for scanning each character 1 at a time
O(N) memory for worst case if the stack is the size of the entire problem
*/

/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {
    let stack = [];
    let dict = {
        ']' : '[',
        '}' : '{',
        ')' : '('
    };

    for(let i = 0; i < s.length; i++){
        let curr_bracket = s[i];
        if((curr_bracket == '[') || (curr_bracket == '{') || (curr_bracket == '(')){
            stack.push(s[i]);
        }
        else{
            if(stack.pop() != dict[curr_bracket]){
                return false;
            }
        }
    }
    if(stack.length > 0){
        return false;
    }
    else{
        return true;
    }
};

File: 20-ValidParenthesis.py
'''
My Solution:

Utilize a stack to keep track of opening bracket sequence order and iterate through the input list one char at a time

O(n) time
O(n) memory for worst case the stack being the entire input list
'''
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        for char in s:

            # if we have an opening bracket, add to the stack
            if char in {'(', '[', '{'}:
                stack.append(char)

            # if we have a closing bracket, pop from the stack and see if the corresponding opening bracket is popped
            else:
                # handle stack is empty edge case
                if not stack:
                    return False
                x = stack.pop()
                if (char == ')' and x != '(') or (char == ']' and x !='[') or (char == '}' and x != '{'):
                    return False

        # only return true if the stack is empty
        if stack:
            return False
        else:
            return True